### DISCUSSION TITLE
Simulator and controller with different timesteps (originally: The simulator's t_step affects the physical system's time constant in the graphs of MPC)

### DISCUSSION THREAD

**Discussion by Yannebeazer** [do-mpc/do-mpc#311] [Q&A] [answered]
**Describe the bug**
Since the simulator's `t_step` is not a property of the simulated physical system, it should not affect the physical system's behaviour. I implemented some tests, however, that suggest that the simulator's `t_step` does affect the physical system's time constant as recorded in the MPC's graphs.

I implemented three simulation runs for the same MPC controlled physical system:

$$
\dot{x} = - x + u
$$

$$
J = x^2 + u^2
$$

To eliminate the MPC's control influence, the input is constrained to $u \ge 0$, such that the MPC always computes $u=0$.

The only difference between these runs is `t_step` of the respective simulator. 

* run_A: (simulator `t_step`: .5 sec.) < (MPC `t_step`: 1 sec.)
* run_B: (simulator `t_step`:  1 sec.) = (MPC `t_step`: 1 sec.)
* run_C: (simulator `t_step`:  2 sec.) > (MPC `t_step`: 1 sec.)

(`run_B` is the Base Case, where the `t_step`s are equal.)

![Screenshot from 2022-08-21 10-38-56](https://user-images.githubusercontent.com/110317249/185770689-eba686a7-8cec-4f66-8694-7c4673309e77.png)

In all cases, the physical system should dampen out in about 4 seconds. This is, however, not the case for the graphs recorded by the MPC in run_A (8 seconds) and in run_C (2 seconds).

**Which version of do-mpc are using and how did you obtain it?**
v4.4.0
- Installation via pip? Via pipenv
- Cloned from repository? No

**Are you running our examples or your own code?**
My own code.
- Is the example version matching your do-mpc version? N/A

**Screenshots**
See above.

**Which system are you working on?**
Linux, Ubuntu 20.04.4 LTS

**Additional context**
For a cooling coil simulation, I like to graph the system dynamics at a lower resolution (say, every 10 seconds) whilst the MPC controls the cooling water flow at a higher resolution (say, every 1 second). In other cases, I could also imagine that someone would like to graph the system dynamics between two MPC control steps at a higher resolution.

**Comment by 4flixt** [comment, marked as answer]
Hi @Yannebeazer,

the ``MPC.data`` object records for the state whatever the MPC controller has received when called with ``mpc.make_step(x0)``. It therefore records the state of the simulator. This state is then assigned to the internal MPC time. At the end of ``mpc.make_step`` this time is advanced by ``t_step``. If you have a different timestep for the simulator, you are therefore assigning the state of a later time. This creates the plotting error.

It is possible to run, e.g. the MPC controller with ``t_step=1`` and the simulator with ``t_step=.5`` by doing something like this:

```python
u = mpc.make_step(x)
for k in range(2):
   x = simulator.make_step(u)
```
which calls the simulator twice for every call of the controller. Vice versa, i.e. calling the controller twice for every result from the simulator, doesnt make sense. You are not obtaining new information, why would you re-compute your optimal control input?

Best,
4flixt

**Reply by Yannebeazer** [reply]
Hi @4flixt,

Thanks for your solution. I think it's worth to  include it in the documentation, since it also illustrates the difference between the two time steps in a clear way.

You are also right about the other way around. I could just as well have skipped the interposing data points from the simulation for the result I was after, which doesn't make sense. 

Sorry for the late reply, by the way, but I just started a new job at a new company that required my full attention.

Kind regards,
Yannebeazer

**Reply by 4flixt** [reply]
All good :) I am happy the answer has helped you. I will keep it in my mind for future updates of the documentation. for now, at least, it is documented through this discussion.

Best,
4flixt
